feat: data set lifecycle job (#588)
FilOzone/dealbot: docs/data-set-termination.md

Reviewer: High level: should we maybe combine this done with data-set-creation.md ? Maybe there's a data-set-lifecycle.md doc we want?

I haven't thought too deeply about this idea, but I'm basically wondering about whether it's easier to think about the dataset lifecycle as a whole rather than spanning two documents.

- If the transaction confirms but `pdpEndEpoch` does not become non-zero before the abort signal fires, treat the run as `failure.timedout` and let pg-boss retry on the next tick.

### Metrics and BetterStack dashboards

The termination job has two distinct observability concerns: is the trigger firing, and is the canary signal it produces showing up in creation metrics. Creation metrics are the primary signal; termination metrics are only there to confirm the trigger is working.

All metrics carry the standard label set defined in [`checks/events-and-metrics.md`](./checks/events-and-metrics.md#metrics):
`checkType`, `providerId`, `providerName`, `providerStatus`.

For termination metrics, `checkType=dataSetTermination`. For creation metrics referenced below, `checkType=dataSetCreation`.

#### Creation metrics (primary signal)

| Metric | `value` labels | What to watch for |
|--------|---------------|-------------------|
| `dataSetCreationStatus`| `pending`, `success`, `failure.timedout`, `failure.other`, `skipped.no_candidate` | `success` count should rise in the interval after each termination; persistent `failure.*` after a termination indicates a `createDataSet` regression |

> | `dataSetCreationMs`| — | Latency histogram for `createDataSetWithPiece`; spikes after termination may indicate on-chain congestion |

Author: [c537509](https://github.com/FilOzone/dealbot/pull/588/commits/c5375095d78b0067638b580d337ceb0ae6a3dcf9)

# Data Set Termination Job

> This doc proposes a calibration-only `data_set_termination` job that periodically terminates a dealbot managed dataset so the existing `data_set_creation` job naturally recreates it. The goal is to keep dealbot continuously exercising the on-chain `createDataSet` lifecycle instead of only creating datasets until a steady-state cap is reached.

Reviewer: Hyperlink to `data_set_creation` docs

> > **Note**: this design does **not** attempt `PDPVerifier.deleteDataSet`, which is SP-initiated; if `deleteDataSet` canary coverage is required for #586, that would need a different approach.

Reviewer: ```suggestion
> **Note**: this design does **not** attempt `PDPVerifier.deleteDataSet`, which is SP-initiated. 
```

This is a good callout, but I think we're agreeing here not do `deleteDataSet` so we don't need to say more.

## Summary

> - `data_set_termination` is a calibration-only job that periodically terminates one managed dataset slot per provider.

Reviewer: I think it should be calibration-only by default, but it should be possible to run in mainnet if someone enables it.

- Together with [`data_set_creation`](./data-set-creation.md), the two jobs form a bounded loop that keeps the `createDataSet` on-chain path continuously exercised as a canary.
- The job terminates **at most one dataset per invocation**; `data_set_creation` handles replenishment on its next scheduled tick.

> - It runs on the shared `sp.work` queue with `singletonKey=spAddress`, so it cannot race with `deal`, `retrieval`, `piece_cleanup`, `pull_check`, or `data_set_creation` for the same provider.

Reviewer: Rather than list them all out (which is going to get stale as we add jobs), maybe just say 

```suggestion
- It runs on the shared `sp.work` queue with `singletonKey=spAddress`, so it cannot race with other jobs for the same provider.
```

- Schedule creation is gated by `NETWORK=calibration` and a non-empty canary window (`MIN_NUM_DATASETS_FOR_CHECKS - DATA_SET_TERMINATION_MIN_INDEX > 0`).
- Slots below `DATA_SET_TERMINATION_MIN_INDEX` are never touched, keeping a stable baseline for ongoing checks.

## Problem Context

> During the PDPVerifier v3.4.0 rollout, `createDataSet` broke on calibration and mainnet. Dealbot did not detect the calibration outage because providers had already reached the steady-state cap of managed datasets ([`MIN_NUM_DATASETS_FOR_CHECKS`](./environment-variables.md#min_num_datasets_for_checks)). Once the cap was reached, `data_set_creation` stopped exercising the on-chain create path, so the canary value of that job disappeared.

Reviewer: ```suggestion
During the PDPVerifier v3.4.0 rollout, `createDataSet` broke on calibration and mainnet. Dealbot did not detect the calibration outage because providers had already reached the steady-state cap of managed datasets ([`MIN_NUM_DATASETS_FOR_CHECKS`](./environment-variables.md#min_num_datasets_for_checks)). Once the cap was reached, `data_set_creation` stopped exercising the on-chain create path, so the canary value of that job disappeared. (See [post mortem](https://app.notion.com/p/filecoindev/2026-05-28-Hotfix-FWSS-v1-2-0-compatibility-with-PDPVerifier-v3-4-0-901dc41950c1820d99c88157214dec5d) for more info.)
```

The missing capability is not more creation logic. The missing capability is a controlled way to create fresh demand for creation again.

## Goals

- Continuously exercise the calibration `createDataSet → terminateService → createDataSet` lifecycle.
- Reuse the existing `data_set_creation` job as the replenishment mechanism.
- Minimize disruption to ongoing deal and retrieval checks.
- Make termination cadence explicitly configurable so the expected create cadence can be reasoned about.

> - Ensure the job cannot run on mainnet.

Reviewer: ```suggestion
- Ensure the job doesn't run on mainnet by default.
```

Introduce a new SP-scoped job type: `data_set_termination`.

The job should:

- run only on calibration
- run on a configurable cadence
- terminate at most one safe managed dataset per provider per invocation
- rely on the existing `data_set_creation` job to recreate the missing slot on a later tick

This keeps termination simple and keeps creation logic centralized in the existing job.

### Configuration

The initial design adds these controls, which follow the same naming pattern as the creation job:

- `DATASET_TERMINATIONS_PER_SP_PER_HOUR`
  - mirrors the existing rate-based job controls
  - converted internally to `intervalSeconds`
  - used to reason about expected termination frequency

- `DATA_SET_TERMINATION_JOB_TIMEOUT_SECONDS`
  - max runtime for one termination job invocation

- `DATA_SET_TERMINATION_MIN_INDEX`
  - the lowest slot index eligible for termination (inclusive)
  - default: `1` — only the baseline slot (index `0`) is protected
  - slots `0` through `DATA_SET_TERMINATION_MIN_INDEX - 1` are never touched by this job
  - example: `MIN_NUM_DATASETS_FOR_CHECKS = 10`, `DATA_SET_TERMINATION_MIN_INDEX = 5` → slots 0–4 are stable, slots 5–9 cycle as the canary window

>   - set to `MIN_NUM_DATASETS_FOR_CHECKS` to disable termination entirely — the canary window becomes empty and no schedule is created

Reviewer: I guess this works, but maybe we use a more obvious "this is disabled" value like "-1" or even a string "DISABLED"?

  - must be `>= 1` and `<=` [`MIN_NUM_DATASETS_FOR_CHECKS`](./environment-variables.md#min_num_datasets_for_checks); violating either constraint crashes the application on startup

### Scheduling and queueing

The scheduling model mirrors `data_set_creation`:

- queue: shared `sp.work`
- `singletonKey=spAddress`

Sharing the singleton with other SP jobs prevents termination from racing with a `deal`, `retrieval`, `pull_check`, `piece_cleanup`, or `data_set_creation` job for the same provider.

The schedule is only upserted when all of the following are true:

> - `NETWORK=calibration`

Reviewer: I don't know if we want to enforce this in code.

Other ideas:
- if the network is mainnet, maybe we log an info/warning (e.g., "Dataset deletion is enabled. Mainnet funds will likely drain faster as a result of ensuing repeated createDataSet fees.") or
- if the network is mainnet, require a more explicit opt in: (e.g., ENABLE_DATASET_DELETION=TRUE") which by default is false for mainnet. Basically I don't want our code to get in the way if we or others decide they want to run deletion jobs against mainnet.

- `MIN_NUM_DATASETS_FOR_CHECKS - DATA_SET_TERMINATION_MIN_INDEX > 0`

The second condition covers the `DATA_SET_TERMINATION_MIN_INDEX = MIN_NUM_DATASETS_FOR_CHECKS` case (empty canary window, termination effectively off) without crashing. It also handles the case where `MIN_NUM_DATASETS_FOR_CHECKS` is later lowered to meet `DATA_SET_TERMINATION_MIN_INDEX` — no schedule is created without requiring a config change.

### Proposed handler algorithm

For one provider, one invocation of `data_set_termination` works like this:

1. Check that the network is calibration. If not, log skip and exit.
2. Apply the same maintenance-window and SP-blocklist rules used by other SP jobs.
3. Create an `AbortController` using `DATA_SET_TERMINATION_JOB_TIMEOUT_SECONDS`.
4. Read `MIN_NUM_DATASETS_FOR_CHECKS` and base dataset metadata.

> 5. Scan slots from `minDataSets - 1` down to `DATA_SET_TERMINATION_MIN_INDEX`. For each slot:

Reviewer: Is the ordering of the slots deterministic? If so, that would mean some datasets will be long lived and some will churn. I haven't thought about this too deeply and appreciate your input, but on the surface, I would imagine doing this with random dataset selection. If we go random, then I assume we do away with `DATA_SET_TERMINATION_MIN_INDEX` approach.

Another way to look at my question is: why are we doing `DATA_SET_TERMINATION_MIN_INDEX` rather than just randomly selecting a dataset?

Author: I thought we keep some slots stable. We don't exercise `dataSetCreation` -> `terminateService` cycle on them because we need deals to perform retrieval tests on them. If we pick random slot from all available slots for terminations we will barely have deals to do retrievals on them.

Author: I was thinking of making only the partition deterministic, not the selection itself. The lower-index slots (up to `DATA_SET_TERMINATION_MIN_INDEX`) stay stable to produce deals for retrieval testing, while the slots in `[DATA_SET_TERMINATION_MIN_INDEX, MIN_NUM_DATASETS_FOR_CHECKS)` can be picked randomly for the `create → terminate` cycle.
I'll make it more clear in the docs.

   - a. Build its metadata using the same logic as `data_set_creation`.
   - b. Classify it via `getDataSetProvisioningStatus()`.
   - c. Skip if `missing` — nothing to terminate.
   - d. Skip if `terminated` — `data_set_creation` owns repair of these slots.
   - e. Skip if `live` but has any deal row with `cleaned_up = false` — the deal job is still tracking it as active.

> 6. Call the termination flow on the first slot that passes all skip conditions (reaches step 5e without being skipped).

Reviewer: ```suggestion
6. Call the termination flow on the first slot that passes all skip conditions (reaches step 5e without being skipped). 
```

7. Log the outcome and exit for this tick.
8. If no eligible slot is found after the full scan, log `skipped.no_candidate` and exit. This is expected when `data_set_creation` has not yet replenished a previously terminated slot.

> As with `data_set_creation`, the job performs **at most one state-changing action per invocation**.

Reviewer: ```suggestion
As with `data_set_creation`, the job performs **at most one state-changing action per invocation**. The way to increase the dataset termination rate is to increase `DATASET_TERMINATIONS_PER_SP_PER_HOUR`.
```

Polling until the chain confirms termination is important because the canary value comes from the full on-chain lifecycle, not just submitting a transaction.

#### Idempotency

The termination flow must tolerate races and retries:

- If `pdpEndEpoch !== 0` is already set when the job starts (slot is already terminated), skip the `terminateService` call and treat the run as a no-op success.
- If `terminateService` reverts with an already-terminated error (for example, `"service already terminated"` or `"dataset not active"`), treat it as idempotent success and proceed to the polling step.
- If the transaction confirms but `pdpEndEpoch` does not become non-zero before the abort signal fires, treat the run as `failure.timedout` and let pg-boss retry on the next tick.

### Metrics and BetterStack dashboards

The termination job has two distinct observability concerns: is the trigger firing, and is the canary signal it produces showing up in creation metrics. Creation metrics are the primary signal; termination metrics are only there to confirm the trigger is working.

> All metrics carry the standard label set defined in [`checks/events-and-metrics.md`](./checks/events-and-metrics.md#metrics):

Reviewer: Do we maybe put all the metrics in events-and-metrics so its all in one place?

- are `dataSetTerminationStatus{value="skipped.no_candidate"}` runs persisting longer than one creation interval, indicating `data_set_creation` is not replenishing?
- does `dataSetCreationStatus{value="success"}` follow `dataSetTerminationStatus{value="success"}` within the expected interval?
- are `dataSetCreationStatus{value="failure.*"}` counts rising after terminations, indicating a regression in `createDataSet`?

## Relationship to `data_set_creation`

The two jobs form a bounded loop.

`data_set_termination` only terminates datasets that correspond to dealbot-managed metadata slots. `data_set_creation` detects the resulting `missing` slot through its normal metadata lookup and recreates it without needing any new cross-job state.

Expected healthy behavior:

1. `data_set_termination` calls `terminateService` and polls until `pdpEndEpoch !== 0`.
2. `data_set_creation` runs next. The Synapse SDK filters the terminated dataset from metadata lookups, so the slot resolves as `missing` immediately. `data_set_creation` provisions a replacement dataset directly in this run.
3. Existing creation metrics and alerts resume acting as the canary.

> **Rate constraint:** `DATASET_CREATIONS_PER_SP_PER_HOUR` should be **greater than or equal to** `DATASET_TERMINATIONS_PER_SP_PER_HOUR`. If termination runs faster than creation, the missing-slot backlog accumulates and the system stops behaving like a simple steady-state canary. The scheduler should emit a startup warning log when this constraint is violated so the misconfiguration is visible without a dashboard.

Reviewer: ```suggestion
**Rate constraint:** `DATASET_CREATIONS_PER_SP_PER_HOUR` should be **greater than or equal to** `DATASET_TERMINATIONS_PER_SP_PER_HOUR`. If termination runs faster than creation, the missing-slot backlog accumulates and the system stops behaving like a simple steady-state canary. The scheduler emits a startup warning log when this constraint is violated so the misconfiguration is visible without a dashboard.
```

I changed the wording because I'm assume we'll implement it this way.

**Canary window size:** The number of slots eligible for termination is `MIN_NUM_DATASETS_FOR_CHECKS - DATA_SET_TERMINATION_MIN_INDEX`. A canary window of `1` means a single slot cycles continuously; a larger window gives termination more candidates when one slot has active deals blocking it. In practice, a window of `2`–`3` is usually enough buffer.

## Open Questions

### Should `terminated` be renamed in `getDataSetProvisioningStatus`?

The `terminated` status returned by `getDataSetProvisioningStatus` means: the Synapse SDK resolved a `dataSetId` from the metadata fingerprint but liveness probes failed. This is distinct from a dataset that has `pdpEndEpoch !== 0` on-chain (which the SDK filters out entirely, causing the slot to resolve as `missing`).

> The name `terminated` is already used for both the on-chain lifecycle concept and this SDK liveness-probe failure state, which causes confusion. Candidate replacements: `irrecoverable` or `missing.sp`. This rename would affect `data_set_creation`'s handler and repair path as well.

Reviewer: I don't know what names are used where and the specifics, but lets not make things confusing. A "terminated" dataset already has a definition (and most recently getting updated in https://app.notion.com/p/filecoindev/Data-Set-Terminations-Clean-up-360dc41950c1801ebf0aff017a322e7 I believe). 

If our terminated definition isn't in sync with other understandings of terminated, then lets not use that word.

I don't really know yet what to suggestion in terms of name.

When you say "liveness probes failed", what do you mean? Basically to help with naming, today when `getDataSetProvisioningStatus` returns `terminated` what is happening with the dataset? How could that have happened to the dataset? That is useful info for coming up with a new name here.

But in general: I'm in favor of fixing this and making it more clear and simpler to reason about. Overlaoding the term "terminated" I think moves us in the wrong direction.

Author: When `getDataSetProvisioningStatus` return `terminated` it means data set is live/healthy on-chain but curio's `addPieces` path returns `unrecoverable_proving_failure_epoch` state, where SP refuses `addPieces` with HTTP 409.

Ref - https://github.com/FilOzone/dealbot/issues/545

> ### Should `data_set_termination` absorb the repair path from `data_set_creation`?

Reviewer: It's good to raise it. I don't know if there is an obvious clear-cut answer.

It sounds like the repair case still involves creating a dataset right, which is why it's reasonable to be in `data_set_creation`.

Maybe we just have an FAQ item for "Why doesn't data-set-termination handle repair?" For the answer we can say it's not clear cut either direction and that logic first lived in create_data_set. There hasn't been a compelling reason to change this.

Author: By repair it means the curio's `addPieces` refuses with `unrecoverable_proving_failure_epoch` but the data set is still live on-chain. So, repair path calls `terminateSerivce` which then actually terminates the data set on-chain. Then, in next tick the data set is replenished.

1. Repairing `terminated` slots (liveness-probe failures) via `repairTerminatedDataSet`.
2. Provisioning `missing` slots via `createDataSetWithPiece`.

Once `data_set_termination` exists and calls `terminateService` directly, it handles on-chain termination for managed slots. The question is whether `data_set_creation` should be simplified to only own replenishment, with all termination (including repair) moving to `data_set_termination`. This is left open pending implementation experience.

## FAQ

### What happens on-chain after `terminateService` is called?

`terminateService` does not delete a dataset instantly. It starts a multi-step on-chain sequence that plays out over roughly 30 days. Understanding this is important because the termination job only needs the first step to complete before it can exit and let `data_set_creation` replenish the slot.

**Step 1 — terminateService tx confirms**

`terminateService` calls `FilecoinPay.terminateRail(pdpRailId)`, which sets `endEpoch = block.number + lockupPeriod` on the PDP rail. The FWSS `railTerminated` callback fires in the same transaction, stores `info.pdpEndEpoch`, and emits `PDPPaymentsTerminated` and `ServiceTerminated`.

> This is the point the termination job polls for: `pdpEndEpoch !== 0`. Once this is set, `data_set_creation` will classify the slot as `missing` and begin the replenishment sequence. The termination job's work is done here.

Reviewer: > termination job polls for: `pdpEndEpoch !== 0`.

Maybe quickly say what happens after this polling when ``pdpEndEpoch !== 0` is found (or you can link to the document where this is described more).

> Once this is set

I think it's not fully obvious what the "this" pronoun is referring to. To make it clear, maybe remove the pronoun and replace with the actual noun?
